Mathlib/Order/Lattice.lean

```diff
@@ -0,0 +1,1424 @@
+/-
+Copyright (c) 2017 Johannes Hölzl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl
+-/
+import Mathlib.Order.Monotone
+import Mathlib.Tactic.Simps.Basic
+--import Mathlib.Tactic.PiInstances
+
+/-!
+# (Semi-)lattices
+
+Semilattices are partially ordered sets with join (greatest lower bound, or `sup`) or
+meet (least upper bound, or `inf`) operations. Lattices are posets that are both
+join-semilattices and meet-semilattices.
+
+Distributive lattices are lattices which satisfy any of four equivalent distributivity properties,
+of `sup` over `inf`, on the left or on the right.
+
+## Main declarations
+
+* `semilattice_sup`: a type class for join semilattices
+* `semilattice_sup.mk'`: an alternative constructor for `semilattice_sup` via proofs that `⊔` is
+  commutative, associative and idempotent.
+* `semilattice_inf`: a type class for meet semilattices
+* `semilattice_sup.mk'`: an alternative constructor for `semilattice_inf` via proofs that `⊓` is
+  commutative, associative and idempotent.
+
+* `lattice`: a type class for lattices
+* `lattice.mk'`: an alternative constructor for `lattice` via profs that `⊔` and `⊓` are
+  commutative, associative and satisfy a pair of "absorption laws".
+
+* `distrib_lattice`: a type class for distributive lattices.
+
+## Notations
+
+* `a ⊔ b`: the supremum or join of `a` and `b`
+* `a ⊓ b`: the infimum or meet of `a` and `b`
+
+## TODO
+
+* (Semi-)lattice homomorphisms
+* Alternative constructors for distributive lattices from the other distributive properties
+
+## Tags
+
+semilattice, lattice
+
+-/
+
+
+universe u v w
+
+variable {α : Type u} {β : Type v}
+
+-- TODO: move this eventually, if we decide to use them
+-- Porting note: no ematch attribute
+--attribute [ematch] le_trans lt_of_le_of_lt lt_of_lt_of_le lt_trans
+
+section
+
+-- TODO: this seems crazy, but it also seems to work reasonably well
+-- Porting note: no ematch attribute
+--@[ematch]
+theorem le_antisymm' [PartialOrder α] : ∀ {a b : α}, a ≤ b → b ≤ a → a = b :=
+  @le_antisymm _ _
+#align le_antisymm' le_antisymm'
+
+end
+
+/-!
+### Join-semilattices
+-/
+
+
+-- TODO: automatic construction of dual definitions / theorems
+/-- A `semilattice_sup` is a join-semilattice, that is, a partial order
+  with a join (a.k.a. lub / least upper bound, sup / supremum) operation
+  `⊔` which is the least element larger than both factors. -/
+-- Porting note: no protect_proj attribute
+--@[protect_proj]
```

Comment:
You can just delete `@[protect_proj]`, but then write `protected ` in front of each field of the class.